doc/dev/schema.md

```diff
@@ -0,0 +1,193 @@
+Depending on return values, it decides to call some reload util and retry
+the function call or pass the error value to the user. Retries are counted
+per function it wraps, so in fact there could be more reloads than ``N``
+for a single request. For example, if a user calls `insert_object`, object
+flatten has failed (and has succeeded after reload) and then storage insert
+has failed and hashes mismatch, reload will happen twice and there will be
+two retries (though it's two different blocks of code will be retried
+independently). Or if a user calls `insert_object_many`, each tuple
+flatten is wrapped independently (though it is hard to imagine a real
+case of multiple flatten retries: user will need to pass a batch of
+tables with different structure and change the schema several times
+to match the current one; in reality it always will be success after
+a single retry or a fast fail).
+
+### Alternative approaches
+
+One of the alternatives considered was to ping a storage instance on
+each request to refresh schema (see [PR#111 comment](https://github.com/tarantool/crud/pull/111#discussion_r562757016)),
+but it was declined due to performance degradation.
+
+
+## Sharding schema
+
+Related links: [#166](https://github.com/tarantool/crud/issues/166),
+[PR#181](https://github.com/tarantool/crud/pull/181),
+[#237](https://github.com/tarantool/crud/issues/237),
+[PR#239](https://github.com/tarantool/crud/pull/239),
+[#212](https://github.com/tarantool/crud/issues/212),
+[PR#268](https://github.com/tarantool/crud/pull/268).
+
+### How schema is stored
+
+Again, ``crud`` cluster is a [``vshard``](https://www.tarantool.io/en/doc/latest/reference/reference_rock/vshard/)
+cluster. Thus, data is sharded based on some key. To extract the key
+from a tuple and compute a [``bucket_id``](https://www.tarantool.io/en/doc/latest/reference/reference_rock/vshard/vshard_architecture/),
+we use [``ddl``](https://github.com/tarantool/ddl) module data.
+``ddl`` module schema describes sharding key (how to extract data
+required to compute ``bucket_id`` from a tuple based on space schema)
+and sharding function (how to compute ``bucket_id`` based on the data).
+This information is stored on storages in some tables and Tarantool
+spaces: ``_ddl_sharding_key`` and ``_ddl_sharding_func``. ``crud``
+module uses `_ddl_sharding_key` and `_ddl_sharding_func` spaces to fetch
+sharding schema: thus, you don't obliged to use ``ddl`` module and can
+setup only ``_ddl_*`` spaces manually, if you want. ``crud`` module uses
+plain Lua tables to store sharding info on routers and storages.
+
+### When schema is used
+
+Sharding schema is used to compute ``bucket_id``. Thus, it is used each
+time we need to execute a non-map-reduce request (``insert_*``,
+``replace_*``, ``update_*``, ``delete_*``,  ``upsert_*``, ``get_*``
+requests, most ``select`` and ``count`` requests, some ``pairs``
+requests), ``bucket_id`` is not specified explicitly and there is no
```

Comment:
If you have a three-line out-of-flow inline comment, it likely will look better as a [footnote](https://github.blog/changelog/2021-09-30-footnotes-now-supported-in-markdown-fields/) rather than as a clause in parentheses.

Reply:
> Thus, it is used each time we need to execute a non-map-reduce request, (?) ``bucket_id`` is not specified explicitly and there is no ``force_map_call`` option.

(Removed the clause in the parentheses.)

I guess something like 'IOW, when' is missed in the place, which I marked with (?).

Reply:
Used a footnote, thank you.